### PR TITLE
feat(fe-fpm-writer): add support to generate building blocks

### PR DESCRIPTION
**Pull request by kranthie-sap**
The following building blocks can be generated using the fe-fpm-writer module:
* [Chart](https://ui5.sap.com/#/api/sap.fe.macros.Chart)
* [FilterBar](https://ui5.sap.com/#/api/sap.fe.macros.FilterBar)
* [Field](https://ui5.sap.com/#/api/sap.fe.macros.Field)

### CODE REVIEW

File: packages/fe-fpm-writer/src/building-block/index.ts [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+    fs: Editor
+): Editor {
+    // Validate the base and view paths
+    validateBasePath(basePath, fs);
+    if (!fs.exists(join(basePath, viewPath))) {
+        throw new Error(`Invalid view path ${viewPath}.`);
+    }
+
+    // Read the view xml and template files and update contents of the view xml file
+    const viewDocument = getViewDocument(basePath, viewPath, fs);
```
Comment (tobiasqueck) [inline packages/fe-fpm-writer/src/building-block/index.ts:34]: Can we also handle fragments?

File: packages/fe-fpm-writer/src/building-block/index.ts [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+
+    return viewDocument;
+}
+/**
+ * Returns the macros namespace from the xml view document.
+ *
+ * @param {Document} viewDocument - the view xml file document
+ * @returns {string} the macros namespace
+ */
+function getMacrosNamespace(viewDocument: Document): string {
```
Comment (tobiasqueck) [inline packages/fe-fpm-writer/src/building-block/index.ts:77]: This is more than just getting it, correct? if you don't find it, you add the namespace. If I understand it correctly, then, I would suggest to change the function name and description to indicate that.

File: packages/fe-fpm-writer/src/building-block/types.ts [hunk 3]
```diff
@@ -0,0 +1,342 @@
...
+
+/**
+ * Represents a Chart building block control.
+ * Usually, a contextPath and metaPath is expected.
+ *
+ * @example
+ * <macro:Chart id="Mychart" contextPath="/RootEntity" metaPath="@com.sap.vocabularies.UI.v1.Chart" />
+ * @extends {BuildingBlock}
+ */
+export interface Chart extends BuildingBlock {
```
Comment (tobiasqueck) [inline packages/fe-fpm-writer/src/building-block/types.ts:45]: Could you please also add the optional event handlers.

**Review by tobiasqueck — COMMENTED** [review]
Code is very clean and easy to understand. I just added a few minor comments, would be great if you can address these. I have not locally tested it yet.
Do you think you can also add one or two building blocks to the integration test? If not, don't worry, we can do it later, however, would be nice to have.

File: packages/fe-fpm-writer/templates/building-block/field/View.xml [hunk 4]
```diff
@@ -0,0 +1,8 @@
+<<%- macrosNamespace %>:Chart
+    id="<%- data.id %>"<% if (data.metaPath) { %>
```
Comment (nlunets) [inline packages/fe-fpm-writer/templates/building-block/field/View.xml:2]: Probably Field rather than Chart :)

**Review by nlunets — CHANGES_REQUESTED** [review]

**Comment by nlunets** [comment]
Overall looks good but one tiny problem with the Field :)

File: packages/fe-fpm-writer/templates/building-block/field/View.xml (hunk 4 above)
Comment (kranthie-sap) [reply packages/fe-fpm-writer/templates/building-block/field/View.xml:2]: Thanks for catching it. :)

File: packages/fe-fpm-writer/src/building-block/types.ts (hunk 3 above)
Comment (kranthie-sap) [reply packages/fe-fpm-writer/src/building-block/types.ts:45]: @tobiasqueck: I see that I missed the `selectionChange` event. I will add this. I haven't added any inherited events, properties or methods. Let me know if you need any other events or properties and I will add them for now. Hopefully, when the types from Fiori elements are used, all properties can be used.

File: packages/fe-fpm-writer/src/building-block/index.ts (hunk 1 above)
Comment (kranthie-sap) [reply packages/fe-fpm-writer/src/building-block/index.ts:34]: The code should work for both xml views and xml fragment files. I updated the public function's variable name and docs to indicate this. I haven't updated this in other places in the file for better readability.

File: packages/fe-fpm-writer/src/building-block/types.ts (hunk 3 above)
Comment (kranthie-sap) [reply packages/fe-fpm-writer/src/building-block/types.ts:45]: I added the `selectionChange` event to the interface and the template.

File: packages/fe-fpm-writer/templates/building-block/field/View.xml (hunk 4 above)
Comment (kranthie-sap) [reply packages/fe-fpm-writer/templates/building-block/field/View.xml:2]: Updated the template and test snapshots.

File: packages/fe-fpm-writer/src/building-block/index.ts (hunk 2 above)
Comment (kranthie-sap) [reply packages/fe-fpm-writer/src/building-block/index.ts:77]: Good catch. Updated the function name.

**Comment by kranthie-sap** [comment]
@tobiasqueck:
> Do you think you can also add one or two building blocks to the integration test? If not, don't worry, we can do it later, however, would be nice to have.

Yes, I plan to update the integration tests in the next PR when I add support for other building blocks.

**Review by nlunets — APPROVED** [review]
All good now :)

File: packages/fe-fpm-writer/src/building-block/types.ts (hunk 3 above)
Comment (tobiasqueck) [reply packages/fe-fpm-writer/src/building-block/types.ts:45]: Thanks, I am ok with it for now, and we can enhance it later.

File: packages/fe-fpm-writer/src/building-block/index.ts [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+ * Generates a building block into the provided xml view file.
+ *
+ * @param {string} basePath - the base path
+ * @param {string} viewOrFragmentPath - the path of the view or fragment xml file relative to the base path
+ * @param {string} aggregationPath - the aggregation xpath
+ * @param {BuildingBlock} buildingBlockData - the building block data
+ * @param {Editor} [fs] - the memfs editor instance
+ * @returns {Editor} the updated memfs editor instance
+ */
+export function generateBuildingBlock<T extends BuildingBlock>(
```
Comment (tobiasqueck) [inline packages/fe-fpm-writer/src/building-block/index.ts:20]: Sorry, I missed this in my first review, however, i would like to have a the `generate` functions consistent in the module. We always have three parameters: `basePath`, an object will all specific configs, and optional `fs`.

File: packages/fe-fpm-writer/src/building-block/index.ts [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+
+/**
+ * Returns the view xml file document.
+ *
+ * @param {string} basePath - the base path
+ * @param {string} viewPath - the path of the xml view relative to the base path
+ * @param {Editor} fs - the memfs editor instance
+ * @returns {Document} the view xml file document
+ */
+function getViewDocument(basePath: string, viewPath: string, fs: Editor): Document {
```
Comment (tobiasqueck) [inline packages/fe-fpm-writer/src/building-block/index.ts:49]: The function also works with fragments, also indicated in your public function, so I'd suggest to rename it as well.

**Review by tobiasqueck — CHANGES_REQUESTED** [review]

**Review by nlunets — APPROVED** [review]

**Review by tobiasqueck — APPROVED** [review]
Provided fixes for my complaints :)

Otherwise, the code looks great, both src and test are well written and easy to understand.
Local tests also worked without issues.

File: packages/fe-fpm-writer/src/building-block/index.ts (hunk 5 above)
Comment (tobiasqueck) [reply packages/fe-fpm-writer/src/building-block/index.ts:20]: https://github.com/SAP/open-ux-tools/pull/555/commits/9229ec75d8eff92ceeca032f392eda1c602d6b5a

**Pull request merged**